File: 1053-previous-permutation-with-one-swap/1053-previous-permutation-with-one-swap.py
class Solution:
    def prevPermOpt1(self, arr: List[int]) -> List[int]:
        n = len(arr); pos = -1;
        for i in range(n-1):
            if(arr[i]>arr[i+1]):
                pos = i
        
        #if arr is sorted pos will remain as -1 so return the original array
        if(pos==-1):
            return arr
        
        pos_to_swap_with = pos+1;
        for j in range(pos+1,n,1):
            if(arr[j]<arr[pos] and arr[j]>arr[pos_to_swap_with]):
                pos_to_swap_with = j
        
        a = arr[pos]
        arr[pos] = arr[pos_to_swap_with]
        arr[pos_to_swap_with] = a
        
        return arr;

File: 1053-previous-permutation-with-one-swap/1053-previous-permutation-with-one-swap.cpp
class Solution {
public:
    vector<int> prevPermOpt1(vector<int>& arr) {
        int pos = -1;
        int  n = arr.size();
        if(n==1) return arr;
        for(int i=0;i<n-1;i++){
            if(arr[i]>arr[i+1]){
                pos = i;
            }
        }
        if(pos==-1) return arr;
        int mini = pos+1;
        for(int j=pos+1;j<n;j++){
            if(arr[j]<arr[pos] && arr[j]>arr[mini]){
                mini = j;
            }
        }
        swap(arr[pos],arr[mini]);
        return arr;
    }
};